apis/projectcontour/v1/httpproxy.go

```diff
@@ -199,6 +199,30 @@ type Service struct {
 	UpstreamValidation *UpstreamValidation `json:"validation,omitempty"`
 	// If Mirror is true the Service will receive a read only mirror of the traffic for this route.
 	Mirror bool `json:"mirror,omitempty"`
+	// RequestHeaders defines how to add or remove headers from requests routed to this service.
+	// +optional
+	RequestHeaders *HeaderAlterations `json:"requestHeaders,omitempty"`
+	// ResponseHeaders defines how to add or remove headers from responses returned from this service.
+	// +optional
+	ResponseHeaders *HeaderAlterations `json:"responseHeaders,omitempty"`
```

Comment:
This should follow the API patter established by `PathRewritePolicy`.

* The struct should be called `HeaderRewritePolicy`
* rewrites should be attached to routes, rather than services

Otherwise, this approach looks good to me

Reply:
That's a pretty interesting use case. One of the rationales for the policy structs is that we can allow them to be applied at different levels. Maybe it makes sense to allow `HeaderRewritePolicy` at both Route and Route.Service levels, though we then need to define how merge works.

Reply:
We started a `HeaderPolicy` design which deals with adding/removing/updating, etc of headers in #1982

Reply:
re: merge, note that the docs you referred me to do talk about the ordering of the levels within Envoy.

re: #1982, sorry if I jumped the gun on this, I wasn't aware of the other issue 😅